### ISSUE TITLE
Automate publishing new package versions and release notes

### ISSUE THREAD

**Issue by chasenlehara** [donejs/donejs#1101]
**TLDR:** use [semantic-release](https://semantic-release.gitbooks.io/semantic-release/) in all our projects to automate publishing to npm, writing release notes, and updating the main `can` package.

> This was discussed on a [recent live stream (21:00)](https://youtu.be/wjBBXb1j9ZU?t=21m00s).

## Problem

As a CanJS user (and hopefully contributor 😃), you’ve probably run into at least one of the following:

1. Going to a project’s Releases page on GitHub and not finding release notes for a new version.
2. If you use the `can` package, then wanting to get the latest bug fixes in a package (like `can-component`) but it not being in a released version of `can`.
3. If you use the split-up packages (like `can-component`), then upgrading to new versions only to find the docs on canjs.com haven’t been updated because a new patch/minor release of `can` hasn’t been made.
4. Contributing a change but not being sure how to write the commit message.
5. When releasing a new version, not being sure which commands to run or if it should be a patch/minor/major release.
6. Needing to get collaborator access on npm to publish new versions.
7. Having to write release notes for new versions… and if you’ve forgotten to, I’ve probably bugged you about it.

As maintainers, I think we would release `can` more frequently if there was less friction to do so; right now we have to make sure every updated package has release notes in order for our generate-release-notes script to be useful in the main `can` repo, and it’s still a manual process to run the release locally, run the release notes script, and post the release notes to GitHub.

With CanJS 5, by default, we’re going to recommend that people use the main `can` package with tree-shaking. If we don’t make new releases in a more timely fashion, people won’t be getting bug fixes as quickly.

I think we can implement [semantic-release](https://semantic-release.gitbooks.io/semantic-release/) to help fix/alleviate all of these problems.

## What is semantic-release?

To briefly summarize what’s in the [semantic-release](https://semantic-release.gitbooks.io/semantic-release/) docs:

- Format commit messages in a particular way to indicate fixes/enhancements/breaking changes; the format is configurable and I think the [eslint](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-eslint) preset might be good for us
- When a new commit lands on `master` (or other branches; I’ll get to that down below), Travis will run semantic-release, which will do the following after a successful build:
	- Look at the new commits to determine if a patch/minor/major version should be published
	- Publish a new release to npm
	- Generate release notes
	- Create a versioned git tag

## How would we use it?

The most simple case is when a PR fixes a bug and we want to release a new version of the package. That would look something like:

- Someone submits a PR with a commit like `Fix: Handle case where… (fixes #15)`
- We review & merge the PR
- Travis runs on `master`; semantic-release publishes new release & updates release notes
- Greenkeeper opens a PR in [canjs/canjs](https://github.com/canjs/canjs) to update the dependency
- We merge the PR to `master`; semantic-release does its magic again and the main `can` package is updated!

For minor and major releases, we can still follow a similar process as what we do now: create `minor` and `major` branches that PRs get merged into while we work on the release, then merge one PR into `master` when the release is ready.

I am a little less sure of how pre-releases should work. There’s at least [one set of plugins](https://github.com/telerik/semantic-prerelease) that support publishing to a different npm dist tag (e.g. `@pre`); that might work for us, or we might do something else, or we might continue to release pre-releases manually (since we usually don’t write release notes for those anyway… not sure if that’s a bug or a feature 😝).

## Alternatives

I’m not familiar with any competitors to semantic-release that meet our needs as well as it does.

If we continue building our own scripts, we would need to:

- Figure out how we want to [automate can-* package release notes](https://github.com/canjs/canjs/issues/3538)
- Update all the [can-* packages to publish to npm via Travis](https://github.com/canjs/canjs/issues/3606)

I think writing and maintaining our own tooling that accomplishes something similar to semantic-release would take us more time for less out-of-the-box flexibility. Additionally, using semantic-release helps contributors who’ve already contributed to other projects that use it.

## What would we need to do to implement it?

For each repo:

- Configure Travis: https://github.com/semantic-release/semantic-release/blob/caribou/docs/recipes/travis.md
- Add Greenkeeper config to create commits that match the format we want: https://greenkeeper.io/docs.html#custom-commit-messages
- Remove (all?) the release scripts

Some of this we could do with codemods, but I think we would need to manually set up some of the required GitHub/npm tokens.

We’d also want to update our contributing guidelines so people know to use a specific format for commit messages.

**Comment by frank-dspeed** [comment]
this is related to https://github.com/canjs/canjs/issues/3606
